File: backend/src/main/java/com/sandbox/service/OrganizationService.java
package com.sandbox.service;

import org.springframework.stereotype.Service;
import java.util.List;

import com.sandbox.dto.OrganizationRequest;
import com.sandbox.entity.Organization;
import com.sandbox.entity.User;
import com.sandbox.repository.OrganizationRepository;
import com.sandbox.exception.ResourceNotFoundException;

@Service
public class OrganizationService {

	private final OrganizationRepository organizationRepository;

	public OrganizationService(OrganizationRepository organizationRepository) {
		this.organizationRepository = organizationRepository;
	}

	public Organization createOrganization(OrganizationRequest request, User currentUser) {

		String domain = request.getDomain();

		if (domain != null && !domain.isBlank()) {
			domain = domain.trim().toLowerCase();

			if (organizationRepository.existsByDomain(domain)) {
				throw new IllegalArgumentException("Organization domain already exists");
			}
		} else {
			domain = null;
		}

		Organization organization = Organization.builder().name(request.getName().trim()).domain(domain)
				.status("ACTIVE").createdBy(currentUser.getId()).build();

		return organizationRepository.save(organization);
	}

	public List<Organization> getAllOrganizations() {
		return organizationRepository.findAll();
	}

	public Organization getOrganizationById(Long id) {
		return organizationRepository.findById(id)
				.orElseThrow(() -> new ResourceNotFoundException("Organization not found"));
	}

	public Organization updateOrganization(Long id, OrganizationRequest request) {

		Organization organization = organizationRepository.findById(id)
				.orElseThrow(() -> new ResourceNotFoundException("Organization not found"));

		String domain = request.getDomain().trim().toLowerCase();

		organizationRepository.findByDomain(domain).filter(existing -> !existing.getId().equals(id))
				.ifPresent(existing -> {
					throw new IllegalArgumentException("Organization domain already exists");
				});

		organization.setName(request.getName().trim());
		organization.setDomain(domain);

		return organizationRepository.save(organization);
	}
	
	public void deleteOrganization(Long id) {

	    Organization organization = organizationRepository.findById(id)
	            .orElseThrow(() ->
	                    new ResourceNotFoundException("Organization not found"));

	    organizationRepository.delete(organization);
	}
	
	public Organization updateOrganizationStatus(
	        Long id,
	        String status) {

	    Organization organization = organizationRepository.findById(id)
	            .orElseThrow(() ->
	                    new ResourceNotFoundException("Organization not found"));

	    organization.setStatus(status);

	    return organizationRepository.save(organization);
	}
}

File: frontend/src/services/organizationService.js
import api from "../api/axios";

/*
|--------------------------------------------------------------------------
| Get all organizations
|--------------------------------------------------------------------------
*/
export const getOrganizations = async () => {
  const response = await api.get("/organizations");
  return response.data;
};

/*
|--------------------------------------------------------------------------
| Get organization by id
|--------------------------------------------------------------------------
*/
export const getOrganizationById = async (id) => {
  const response = await api.get(`/organizations/${id}`);
  return response.data;
};

/*
|--------------------------------------------------------------------------
| Create organization
|--------------------------------------------------------------------------
*/
export const createOrganization = async (organization) => {
  const response = await api.post("/organizations", organization);
  return response.data;
};

/*
|--------------------------------------------------------------------------
| Update organization
|--------------------------------------------------------------------------
*/
export const updateOrganization = async (id, organization) => {
  const response = await api.put(`/organizations/${id}`, organization);
  return response.data;
};

/*
|--------------------------------------------------------------------------
| Change organization status
|--------------------------------------------------------------------------
*/
export const updateOrganizationStatus = async (id, status) => {
  const response = await api.patch(`/organizations/${id}/status`, {
    status,
  });
  return response.data;
};

/*
|--------------------------------------------------------------------------
| Delete organization
|--------------------------------------------------------------------------
*/
export const deleteOrganization = async (id) => {
  const response = await api.delete(`/organizations/${id}`);
  return response.data;
};